service-base/src/test/java/org/eclipse/hono/service/amqp/AmqpServiceBaseTest.java

```diff
@@ -0,0 +1,172 @@
+                return ENDPOINT;
+            }
+
+            @Override
+            public void onLinkAttach(final ProtonConnection con, final ProtonReceiver receiver, final ResourceIdentifier targetResource) {
+                linkEstablished.countDown();
+            }
+
+            @Override
+            protected boolean passesFormalVerification(ResourceIdentifier targetAddress, Message message) {
+                return true;
+            }
+        };
+        AuthorizationService authService = mock(AuthorizationService.class);
+        when(authService.isAuthorized(Constants.PRINCIPAL_ANONYMOUS, targetAddress, Activity.WRITE)).thenReturn(Future.succeededFuture(Boolean.TRUE));
+        AmqpServiceBase<ServiceConfigProperties> server = createServer(endpoint);
+        server.setAuthorizationService(authService);
+
+        // WHEN a client connects to the server using this endpoint
+        final Target target = getTarget(targetAddress);
+        final ProtonReceiver receiver = mock(ProtonReceiver.class);
+        when(receiver.getRemoteTarget()).thenReturn(target);
+        when(receiver.attachments()).thenReturn(mock(Record.class));
+        handleReceiverOpen(server, newConnection(Constants.PRINCIPAL_ANONYMOUS), receiver);
+
+        // THEN the server delegates link establishment to the endpoint
+        assertTrue(linkEstablished.await(1, TimeUnit.SECONDS));
+    }
+
+    @Test
+    public void testHandleReceiverOpenRejectsUnauthorizedClient() throws InterruptedException {
+
+        // GIVEN a server with a telemetry endpoint
+        final ResourceIdentifier restrictedTargetAddress = ResourceIdentifier.from(ENDPOINT, "RESTRICTED_TENANT", null);
+        final Endpoint endpoint = mock(Endpoint.class);
+        when(endpoint.getName()).thenReturn(ENDPOINT);
+        AuthorizationService authService = mock(AuthorizationService.class);
+        when(authService.isAuthorized(Constants.PRINCIPAL_ANONYMOUS, restrictedTargetAddress, Activity.WRITE)).thenReturn(Future.succeededFuture(Boolean.FALSE));
+        AmqpServiceBase<ServiceConfigProperties> server = createServer(endpoint);
+        server.setAuthorizationService(authService);
+
+        // WHEN a client connects to the server using a telemetry address for a tenant it is not authorized to write to
+        final CountDownLatch linkClosed = new CountDownLatch(1);
+        final Target target = getTarget(restrictedTargetAddress);
+        final ProtonReceiver receiver = mock(ProtonReceiver.class);
+        when(receiver.getRemoteTarget()).thenReturn(target);
+        when(receiver.setCondition(any())).thenReturn(receiver);
+        when(receiver.close()).thenAnswer(invocation -> {
+            linkClosed.countDown();
+            return receiver;
+        });
+        server.handleReceiverOpen(newConnection(Constants.PRINCIPAL_ANONYMOUS), receiver);
+
+        // THEN the server closes the link with the client
+        assertTrue(linkClosed.await(1, TimeUnit.SECONDS));
```

Comment:
in the meantime I have learned to value using the vertx `TestContext` `Async` for doing these kind of things, because it doesn't require us to throw an `InterruptedException` ...